netket/operator/_pauli_strings/jax.py

```diff
@@ -0,0 +1,355 @@
+    ]
+
+    x_flip_masks_stacked = jnp.concatenate(x_flip_masks, axis=0)
+    z_data = (weights, z_sign_masks, z_sign_indices, z_sign_indices_masks)
+    return x_flip_masks_stacked, z_data
+
+
+@jax.jit
+def _pauli_strings_mels_jax(local_states, z_data, x):
+    # supports both masks and indexing (can be padded, so also with a mask but smaller)
+    # which path is taken is flexible, and can be fully determined by z_data
+    # z_data: a list of tuples weights, z_sign_mask, z_sign_indices, z_sign_indexmask
+    state1 = local_states[1]
+    was_state_1 = x == state1
+    mels = []
+    for w, z_sign_mask, z_sign_indices, z_sign_indexmask in zip(*z_data):
+        if z_sign_mask is not None:  # use masks
+            was_state = was_state_1[..., None, None, :]
+            mask = z_sign_mask
+        else:  # use indexing
+            assert z_sign_indices is not None
+            was_state = x[..., z_sign_indices] == state1
+            if z_sign_indexmask is None:  # no padding, so no mask necessary
+                mask = 1
+            else:
+                mask = z_sign_indexmask
+        # sgn = (-1)**(was_state*mask).sum(axis=-1)
+        sgn = (1 - 2 * (was_state * mask).astype(np.int8)).prod(
+            axis=-1, promote_integers=False
+        )
+        mels.append(jnp.einsum("...ab,ab->...a", sgn, w))
+        return jnp.concatenate(mels, axis=-1)
+
+
+@jax.jit
+def _pauli_strings_kernel_jax(local_states, x_flip_masks_all, z_data, x):
+    # can re-use function from Ising
+    x_prime = _ising_conn_states_jax(x[..., None, :], x_flip_masks_all, local_states)
+    mels = _pauli_strings_mels_jax(local_states, z_data, x)
+    # TODO do we want a cutoff?
+    return x_prime, mels
+
+
+@register_pytree_node_class
+class PauliStringsJax(PauliStringsBase, DiscreteJaxOperator):
+    """
+    Jax-compatible version of :class:`netket.operator.PauliStrings`.
+    """
+
+    @wraps(PauliStringsBase.__init__)
+    def __init__(
+        self,
+        hilbert: AbstractHilbert,
+        operators: Union[str, List[str]] = None,
+        weights: Union[float, complex, List[Union[float, complex]]] = None,
+        *,
+        cutoff: float = 0.0,
+        dtype: DType = complex,
+    ):
+        super().__init__(hilbert, operators, weights, cutoff=cutoff, dtype=dtype)
+
+        if len(self.hilbert.local_states) != 2:
+            raise ValueError(
+                "PauliStringsJax only supports Hamiltonians with two local states"
+            )
+
+        # check that it is homogeneous, throw error if it's not
+        if not isinstance(self.hilbert, HomogeneousHilbert):
+            local_states = self.hilbert.states_at_index(0)
+            if not all(
+                np.allclose(local_states, self.hilbert.states_at_index(i))
+                for i in range(self.hilbert.size)
+            ):
+                raise ValueError(
+                    "Hilbert spaces with non homogeneous local_states are not "
```

Comment:
does it make sense, generally speaking, to have non-homogenous hilbert spaces in this context? not sure it will be ever supported, right?

Reply:
We could eventually support it but it does not really make sense, indeed.